keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -2049,6 +2046,247 @@ Unreserve is called in two scenarios:
+new "ResourcePlugin" type will be used in the Type field of the
+[PluginInfo](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/pluginregistration/v1#PluginInfo)
+response to distinguish the plugin from device and CSI plugins.
+
+Under the advertised Unix Domain socket the kubelet plugin provides the
+k8s.io/kubelet/pkg/apis/dra gRPC interface. It was inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts removed.
+
+#### REST proxy
+
+Previously, kubelet retrieved ResourceClaims and published ResourceSlices on
+behalf of DRA drivers on the node. The information included in those got passed
+between API server, kubelet, and kubelet plugin using the version of the
+resource.k8s.io used by the kubelet. Combining a kubelet using some older API
+version with a plugin using a new version was not possible because conversion
+of the resource.k8s.io types is only supported in the API server and an old
+kubelet wouldn't know about a new version anyway.
+
+Keeping kubelet at some old release while upgrading the control and DRA drivers
+is desirable and officially supported by Kubernetes. To support the same when
+using DRA, the kubelet now provides a REST proxy via gRPC that can be used by
+drivers to execute HTTP requests against the API server. The drivers determine
+the version of the resource.k8s.io. In those few cases where the kubelet needs
+to access the resource.k8s.io API itself, it does so with a dynamic client.
+
+The alternative to implementing a generic REST proxy would have been to create
+dedicated gRPC APIs for all operations that are needed by a driver. This is a
+larger API and prevents using the normal client-go APIs. The way it is
+implemented now, client-go in the plugin is instantiated with a HTTP
+roundtripper provided by
+k8s.io/dynamic-resource-allocation/restproxy. Implementations of that helper
+code in other languages than Go are possible and may provide similar benefits.
+
+##### Security
+
+Requests originating from a driver will be sent to the API server with
+credentials of the kubelet. This has the advantage that the node authorizer can
+be used to limit access to objects belonging to the node. The downside is that
+a driver might abuse this to access some other resources that kubelet has
+access to, like pods. To prevent this, the REST proxy filters requests by path
+and method and only passes requests through which the driver is meant to have
+access to (ResourceSlice and ResourceClaim). Access to ResourceClaims is further
+limited to read-only access by the node authorizer.
+
+In addition, the REST proxy adds `nodeName` and `driverName` field selectors to
+the header of requests that list ResourceSlice objects. This is mostly for
+convenience because for a PUT of a ResourceSlice, the driver has to be trusted
+to not create an object for some other driver on the node. This cannot be
```

Comment:
This is a downside of the proxy path, right? The unstructured / dynamic client aggregation approach could have checked the driver association field, right?

Reply:
Correct.

However, the "unstructured / dynamic client" then has other downsides. Right now, it can check this because we assume that all future ResourceSlices will be of the form "nodeName + driverName + other fields". Without knowing anything about "other fields", it's hard to write a controller that synchronizes the content in the ResourceSlice objects with the desired content - not impossible, but harder.

The other downside if we consider apiserver traffic is the usage of JSON for the request bodies (both ways) instead of protobuf.

Reply:
Why do we want kubelet in the middle here if drivers are going to be doing more than write-only status reporting? If they are watching / getting / reading / reconciling status, that seems more like a normal client to me. Is the only reason for the node proxy so they can piggy-back on kubelet credentials and get authorized?

Reply:
The advantage is indeed that the node authorizer continues to work.

Reply:
Drivers on the same node have to trust each other. They often run with elevated privileges, so they can already do much harm locally even without this shared access to ResourceSlice objects of the node.

That's still better than having to trust all drivers anywhere in the cluster.

Reply:
I feel like this is the ideal scenario for us to build out the authz capabilities of node bound SA tokens (possibly combined with https://github.com/kubernetes/enhancements/pull/4600).  We have a good story for node agents that want to perform write requests via VAP+SA node ref.  Having something similar to that for reads seems broadly useful.  Building an entire gRPC proxy as a one-off for DRA seems like the wrong approach.

Reply:
@pohly I added this to the SIG Auth meeting for May 22nd.  Hopefully we can hash out a concrete path forward.

Reply:
It gets even more interesting with some uses cases that people have brought up for DRA: a DRA driver for a NIC runs on the node and must publish network-related information, ideally to the ResourceClaim status. It must do that itself because kubelet might not know what that information is due to version skew.

The node authorizer can limit write access to ResourceClaims that are actually in use on the node because it builds the graph (node -> pod -> claim). In contrast to ResourceSlice, there isn't a specific field in a ResourceClaim that can be checked.